analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs

```diff
@@ -18,18 +18,22 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using System.Collections.Concurrent;
 using SonarAnalyzer.SymbolicExecution.Constraints;
 
 namespace SonarAnalyzer.SymbolicExecution.Roslyn
 {
     public sealed record SymbolicValue
     {
+        private static ConcurrentDictionary<ConstraintKind, SymbolicValue> singleConstraintCache = new();
```

Comment:
The implementation is almost the same. see commit ee12c3e for the change needed.

I did some performance measurements and the difference with concurrent access is neglectable. For `SymbolicValue.Null.WithConstraint(ObjectConstraint.NotNull)` was called 10.000 times from concurrent threads (via `Parallel.For`).
* Uncached: original implementation
* Cached: Using the ConcurrentDictionary
* Using the switch instead of the ConcurrentDictionary

with these results:

| Method                       | Job      | Runtime   | Allocated  |
|------------------------------|----------|-----------|------------|
|   Null_to_NotNull_Uncached   | .NET 7.0 | 271.93 us | 1254.46 KB |
| Null_to_NotNull_Cached       | .NET 7.0 | 84.12 us  | 4.48 KB    |
| Null_to_NotNull_ObjectCached | .NET 7.0 | 66.51 us  | 4.47 KB    |

Full results for different "WithContraint" and "WithoutConstraint" calls for .Net7 and Net48:

|                                         Method |                Job |            Runtime |        Mean |     Error |    StdDev |      Median | Ratio | RatioSD |     Gen0 |    Gen1 |  Allocated | Alloc Ratio |
|----------------------------------------------- |------------------- |------------------- |------------:|----------:|----------:|------------:|------:|--------:|---------:|--------:|-----------:|------------:|
|                         Empty_to_Null_Uncached |           .NET 7.0 |           .NET 7.0 |   177.92 us |  2.304 us |  2.042 us |   177.80 us |  0.43 |    0.00 | 102.5391 |  2.1973 | 1254.52 KB |        0.99 |
|                         Empty_to_Null_Uncached | .NET Framework 4.8 | .NET Framework 4.8 |   411.68 us |  2.380 us |  2.110 us |   411.20 us |  1.00 |    0.00 | 205.0781 |  2.9297 | 1262.92 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                           Empty_to_Null_Cached |           .NET 7.0 |           .NET 7.0 |    43.30 us |  0.391 us |  0.327 us |    43.38 us |  0.44 |    0.01 |   0.3662 |       - |    4.48 KB |        1.49 |
|                           Empty_to_Null_Cached | .NET Framework 4.8 | .NET Framework 4.8 |    97.76 us |  0.925 us |  0.865 us |    98.19 us |  1.00 |    0.00 |   0.3662 |       - |       3 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                     Empty_to_Null_ObjectCached |           .NET 7.0 |           .NET 7.0 |    35.22 us |  0.678 us |  0.601 us |    35.00 us |  0.42 |    0.01 |   0.3662 |       - |    4.48 KB |        1.49 |
|                     Empty_to_Null_ObjectCached | .NET Framework 4.8 | .NET Framework 4.8 |    83.84 us |  1.200 us |  1.122 us |    83.26 us |  1.00 |    0.00 |   0.4883 |       - |    3.02 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                       Null_to_NotNull_Uncached |           .NET 7.0 |           .NET 7.0 |   271.93 us |  1.801 us |  1.596 us |   271.70 us |  0.52 |    0.00 | 102.5391 |  1.9531 | 1254.46 KB |        0.99 |
|                       Null_to_NotNull_Uncached | .NET Framework 4.8 | .NET Framework 4.8 |   525.38 us |  2.730 us |  2.280 us |   525.38 us |  1.00 |    0.00 | 205.0781 |  2.9297 | 1262.57 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                         Null_to_NotNull_Cached |           .NET 7.0 |           .NET 7.0 |    84.12 us |  0.610 us |  0.509 us |    84.25 us |  0.46 |    0.00 |   0.3662 |       - |    4.48 KB |        1.51 |
|                         Null_to_NotNull_Cached | .NET Framework 4.8 | .NET Framework 4.8 |   184.27 us |  0.886 us |  0.786 us |   184.10 us |  1.00 |    0.00 |   0.2441 |       - |    2.97 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                   Null_to_NotNull_ObjectCached |           .NET 7.0 |           .NET 7.0 |    66.51 us |  0.322 us |  0.301 us |    66.53 us |  0.39 |    0.00 |   0.3662 |       - |    4.47 KB |        1.50 |
|                   Null_to_NotNull_ObjectCached | .NET Framework 4.8 | .NET Framework 4.8 |   170.86 us |  0.902 us |  0.753 us |   171.04 us |  1.00 |    0.00 |   0.2441 |       - |    2.99 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                          Null_to_Null_Uncached |           .NET 7.0 |           .NET 7.0 |    39.82 us |  0.206 us |  0.183 us |    39.85 us |  0.39 |    0.01 |   0.3662 |       - |    4.48 KB |        1.50 |
|                          Null_to_Null_Uncached | .NET Framework 4.8 | .NET Framework 4.8 |   101.50 us |  2.021 us |  4.036 us |    99.40 us |  1.00 |    0.00 |   0.3662 |       - |       3 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                            Null_to_Null_Cached |           .NET 7.0 |           .NET 7.0 |    43.40 us |  0.220 us |  0.206 us |    43.43 us |  0.44 |    0.00 |   0.3662 |       - |    4.47 KB |        1.49 |
|                            Null_to_Null_Cached | .NET Framework 4.8 | .NET Framework 4.8 |    99.24 us |  1.074 us |  0.952 us |    99.01 us |  1.00 |    0.00 |   0.3662 |       - |       3 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                      Null_to_Null_ObjectCached |           .NET 7.0 |           .NET 7.0 |    42.91 us |  0.262 us |  0.232 us |    42.82 us |  0.42 |    0.01 |   0.3662 |       - |    4.47 KB |        1.50 |
|                      Null_to_Null_ObjectCached | .NET Framework 4.8 | .NET Framework 4.8 |   102.21 us |  1.363 us |  1.209 us |   102.42 us |  1.00 |    0.00 |   0.3662 |       - |    2.98 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                  Empty_to_NotNullTrue_Uncached |           .NET 7.0 |           .NET 7.0 |   484.52 us |  1.183 us |  1.107 us |   484.59 us |  0.51 |    0.00 | 255.8594 |  6.8359 | 3129.65 KB |        0.99 |
|                  Empty_to_NotNullTrue_Uncached | .NET Framework 4.8 | .NET Framework 4.8 |   958.83 us |  3.294 us |  2.750 us |   958.26 us |  1.00 |    0.00 | 511.7188 | 10.7422 | 3149.11 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                    Empty_to_NotNullTrue_Cached |           .NET 7.0 |           .NET 7.0 |   337.23 us |  1.824 us |  1.523 us |   337.47 us |  0.47 |    0.00 | 153.8086 |  2.9297 | 1879.58 KB |        0.99 |
|                    Empty_to_NotNullTrue_Cached | .NET Framework 4.8 | .NET Framework 4.8 |   715.16 us |  3.419 us |  2.669 us |   715.30 us |  1.00 |    0.00 | 307.6172 |  4.8828 | 1890.57 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|              Empty_to_NotNullTrue_ObjectCached |           .NET 7.0 |           .NET 7.0 |   325.58 us |  4.061 us |  3.798 us |   326.57 us |  0.47 |    0.01 | 153.8086 |  2.9297 | 1879.41 KB |        0.99 |
|              Empty_to_NotNullTrue_ObjectCached | .NET Framework 4.8 | .NET Framework 4.8 |   686.57 us |  6.531 us |  5.790 us |   686.16 us |  1.00 |    0.00 | 307.6172 |  4.8828 | 1890.57 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                      NotNull_to_Empty_Uncached |           .NET 7.0 |           .NET 7.0 |   117.13 us |  0.568 us |  0.443 us |   117.01 us |  0.47 |    0.00 |  19.5313 |  0.2441 |  238.72 KB |        1.00 |
|                      NotNull_to_Empty_Uncached | .NET Framework 4.8 | .NET Framework 4.8 |   246.86 us |  1.824 us |  1.617 us |   246.28 us |  1.00 |    0.00 |  38.5742 |       - |  238.97 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                        NotNull_to_Empty_Cached |           .NET 7.0 |           .NET 7.0 |    53.84 us |  1.070 us |  2.414 us |    53.09 us |  0.45 |    0.02 |   0.3052 |       - |    4.33 KB |        1.45 |
|                        NotNull_to_Empty_Cached | .NET Framework 4.8 | .NET Framework 4.8 |   119.55 us |  0.860 us |  0.804 us |   119.12 us |  1.00 |    0.00 |   0.3662 |       - |    2.99 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|                  NotNull_to_Empty_ObjectCached |           .NET 7.0 |           .NET 7.0 |    51.16 us |  0.184 us |  0.154 us |    51.14 us |  0.43 |    0.00 |   0.3662 |       - |    4.48 KB |        1.50 |
|                  NotNull_to_Empty_ObjectCached | .NET Framework 4.8 | .NET Framework 4.8 |   119.93 us |  1.106 us |  1.034 us |   119.87 us |  1.00 |    0.00 |   0.3662 |       - |    2.99 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|     NotNull_to_TrueNotNull_to_NotNull_Uncached |           .NET 7.0 |           .NET 7.0 |   523.35 us |  2.467 us |  2.060 us |   523.28 us |  0.51 |    0.00 | 255.8594 |  5.8594 | 3129.67 KB |        0.99 |
|     NotNull_to_TrueNotNull_to_NotNull_Uncached | .NET Framework 4.8 | .NET Framework 4.8 | 1,017.69 us |  5.413 us |  5.064 us | 1,019.47 us |  1.00 |    0.00 | 511.7188 | 11.7188 | 3149.22 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
|       NotNull_to_TrueNotNull_to_NotNull_Cached |           .NET 7.0 |           .NET 7.0 |   572.70 us | 10.948 us | 10.753 us |   570.23 us |  0.37 |    0.01 | 153.3203 |  2.9297 | 1879.54 KB |        0.99 |
|       NotNull_to_TrueNotNull_to_NotNull_Cached | .NET Framework 4.8 | .NET Framework 4.8 | 1,561.66 us |  6.046 us |  5.655 us | 1,559.96 us |  1.00 |    0.00 | 306.6406 |  3.9063 | 1890.66 KB |        1.00 |
|                                                |                    |                    |             |           |           |             |       |         |          |         |            |             |
| NotNull_to_TrueNotNull_to_NotNull_ObjectCached |           .NET 7.0 |           .NET 7.0 |   562.32 us |  7.111 us |  6.304 us |   561.33 us |  0.36 |    0.01 | 153.3203 |  2.9297 | 1879.55 KB |        0.99 |
| NotNull_to_TrueNotNull_to_NotNull_ObjectCached | .NET Framework 4.8 | .NET Framework 4.8 | 1,544.69 us | 11.270 us |  9.991 us | 1,543.65 us |  1.00 |    0.00 | 306.6406 |  3.9063 | 1890.83 KB |        1.00 |

Reply:
Ok then, I'll keep this thread open for archeology / future reference